data_structures/rtree.cpp

```diff
@@ -111,23 +127,131 @@ bool spatial::Rtree<T>::Node::insert(index_t const point_idx) {
 template<typename T>
 void spatial::Rtree<T>::Node::split(int const branch_idx) {
     // pop the overflowing branch from the 'entries' vector
+    auto const overflowing_node = entries[branch_idx].get_node();
+    entries.erase(entries.begin() + branch_idx);
 
-    // pick some seed entries using the split heuristic
+    // make some seed entries using the split heuristic
+    pick_seeds(overflowing_node->entries);
 
-    // push the newly seeded branches onto 'entries'
+    // distribute the leftover child entries between the seeds
+    distribute(overflowing_node->entries);
+}
 
-    // distribute the leftover points
+/**
+ * Pick a number of "good" seed MBBs from the given choices.
+ * Currently, this function uses the quadratic split heuristic.
+ */
+template<typename T>
+void spatial::Rtree<T>::Node::pick_seeds(
+    std::vector<Entry> const& entry_choices
+) {
+    unsigned best_e1 = -1, best_e2 = -1;
+    area_t max_d = -1;
+    for (unsigned i=0; i<entry_choices.size(); i++) {
+        Entry const e1 = entry_choices[i];
+        for (unsigned j=i+1; j<entry_choices.size(); j++) {
+            Entry const e2 = entry_choices[j];
+            Rectangle const r = min_bounding_box(e1.get_mbb(), e2.get_mbb());
+            area_t const d = area(r) - area(e1.get_mbb()) - area(e2.get_mbb());
+            if (d > max_d) {
+                max_d = d;
+                best_e1 = i;
+                best_e2 = j;
+            }
+        }
+    }
+    // create two new entries corresponding to the above seed MBBs
+    entries.push_back(Entry(
+        entry_choices[best_e1].get_mbb(),
+        std::make_shared<Node>()
+    ));
+    entries.push_back(Entry(
+        entry_choices[best_e2].get_mbb(),
+        std::make_shared<Node>()
+    ));
 }
 
 /**
- * When the root node overflows, we need some special logic, 
- * since it has no parent node.
+ * Distribute leftover entries after splitting an overflowing node.
+ * This function is a bit of a mouthful, particularly the if/else stack
+ * at the end, so it could probably do with some revisiting.
  */
 template<typename T>
-void spatial::Rtree<T>::split_root() {
-    // make a new root, with the old root being its only entry
+void spatial::Rtree<T>::Node::distribute(
+    std::vector<Entry>& leftover_entries
```

Comment:
If you'd like to read more about `std::move` and x-values, this is a good example of where it is applicable. You could rewrite this signature as:
```
template < typename T >
void spatial::Rtree<T>::distribute(  std::vector< Entry > && leftover_entries )
```
Observe the double-ampersand. This indicates that the calling function will no longer use these values and this function is free to destroy it. It signals ownership, which is a key part of understanding smart pointers. However, this is really a digression... :smile: